File: Backend/src/services/RolesServices.ts
import { any } from "zod";
import { Roles } from "../domain";
import { NameTables } from "../domain/Enums/NameTables";
import { IRolesServices } from "./interfaces";
import { RolesRepository } from "./repositories";


class RolesServices implements IRolesServices{
    private roleRepository : RolesRepository
    constructor(){
        this.roleRepository = new RolesRepository();
    }
    async GetRoleByRoleName(roleName: string): Promise<Roles> {
        let result: any;
        try{
            result = await this.roleRepository.findBy(NameTables.ROLES, 'roleName', roleName);
            if(result !== undefined){
                const role = new Roles({
                    id: result.id,
                    roleName: result.roleName,
                    typeid: result.typeid
                })
                return role
            }
            return result
        }catch(err){
            throw new Error(`${err}`);
        }
        throw new Error("Method not implemented.");
    }
    async GetbyId(id: number): Promise<Roles> {
        try{
            const result: any = await this.roleRepository.GetbyId(NameTables.ROLES, id);
            if(result !== undefined){
                const role = new Roles({
                    id: result.id,
                    roleName: result.roleName,
                    typeid: result.typeid
                })
                return role
            }
            return result
        }catch(err){
            throw new Error(`${err}`)
        }
    }
    async GetAll(): Promise<Roles[]> {
        let resultRoles: any[] = [];
        let listRoles: Roles[] = [];
        try{
            resultRoles =  await this.roleRepository.listAll(NameTables.ROLES);
            resultRoles.map(Dbroles => {
                const roles = new Roles({
                    id: Dbroles.id,
                    roleName: Dbroles.roleName,
                    typeid: Dbroles.typeid
                });
                listRoles.push(roles);
            })
            return listRoles
        }catch(err){
            throw new Error(`${err}`)
        }
    }
    async Save(entity: Roles): Promise<string> {
        let result: string = ""
        try{
            if(entity.id === 0){
                await this.roleRepository.save(entity)
                result = 'Salvo com sucesso!'
            }else{
                await this.roleRepository.update(entity)
                result = 'Atualizado com sucesso!'
            }
            return result
        }catch(err){
            return `${err}`
        }
    }

}
export { RolesServices }

File: Back/src/services/RolesServices.py
from services.Repositories.RolesRepository import RolesRepository;
from domain.Entities.Roles import Roles;
from domain.Enums.Tables import Tables;


class RolesServices:
    def __init__(self) -> None:
        self.rolesRepository = RolesRepository();

    async def Save(self, entity: Roles):
        try:
            if(entity.id == 0):
                await self.rolesRepository.Save(entity);
                return "Função salva com sucesso!";
            else:
                await self.rolesRepository.Update(entity);
                return "Função atualiza com sucesso!";
        except Exception as e:
            print(e)
    async def VerifyRoleNameExists(self,roleName: str):
        result = await self.rolesRepository.CheckPerProperty(Tables.ROLES.value, 'roleName', roleName);
        if result == True:
            return True;
        return False;
    async def GetById(self, id: int):
        try:
            busca = await self.rolesRepository.FindById(Tables.ROLES.value,id);
            if busca != None:
                for item in busca:
                    role = Roles(item['id'], item['roleName'], item['typeid']);
                return role
            return busca;
        except Exception as ex:
            print("Error :",ex);
    async def GetRoleByRoleName(self, roleName: str):
        try:
            result: any;
            result = await self.rolesRepository.FindBy(Tables.ROLES.value, 'roleName', roleName);
            if result != None:
                role = Roles(result['id'], result['roleName'],result['typeid']);
                return role;
            return None;
        except Exception as e:
            print(e);
    async def Delete(self, id):
        try:
            await self.rolesRepository.Delete(Tables.ROLES.value, id);
            return "Deletado com sucesso!"
        except Exception as e:
            print(e)
    async def GetAll(self):
        try:
            listRoles:list[Roles]= []
            lista = await self.rolesRepository.List(Tables.ROLES.value);
            if lista.__len__() > 0:
                for item in lista:
                    role = Roles(item['id'],item['roleName'],item['typeid']);
                    listRoles.append(role);
                return listRoles;
            return listRoles;
        except Exception as ex:
            print(ex)

    async def CheckExistsRole(self, role: Roles):
        try:
            check = await self.rolesRepository.CheckExistsEntityWithId(role);
            return check;
        except Exception as ex:
            print(ex);

    async def CheckPropertyRoleExists(self, role:Roles):
        try:
            exists = await self.rolesRepository.CheckExistsEntity(role);
            return exists;
        except Exception as ex:
            print(ex);
